util/coverage/coverage_off_target.py

```diff
@@ -0,0 +1,253 @@
+
+    Returns:
+        Stdout lines in a list. Empty lines are filtered out.
+    """
+    log.debug(f"command: {' '.join(args)}")
+    try:
+        res = subprocess.run(args,
+                             stdout=subprocess.PIPE,
+                             stderr=subprocess.PIPE,
+                             encoding="utf-8",
+                             check=True)
+    except subprocess.CalledProcessError as e:
+        log.error(f"stdout: {e.stdout if e.stdout else '(empty)'}")
+        log.error(f"stderr: {e.stderr if e.stderr else '(empty)'}")
+        raise
+    log.debug(f"stdout: {res.stdout if res.stdout else '(empty)'}")
+    log.debug(f"stderr: {res.stderr if res.stderr else '(empty)'}")
+    return [line for line in res.stdout.splitlines() if line]
+
+
+def gather_and_build_libs():
+    """Find all device libraries that can be built on the host build them with coverage
+    instrumentation.
+
+    Returns:
+        The list of device libraries that can be built on the host.
+    """
+    # Query for device libraries to be instrumented.
+    DEVICE_LIBS_INC = [
+        "//sw/device/silicon_creator/...",
+        "//sw/device/lib/...",
+    ]
+    DEVICE_LIBS_EXC = [
+        "//sw/device/lib/dif/...",
+        "//sw/device/lib/testing/... ",
+        "//sw/device/lib/runtime/...",
+        "//sw/device/lib/crypto/...",
+        "//sw/device/lib/arch/...",
+        "//sw/device/lib/ujson/...",
+        "//sw/device/lib/base:status",
+        "//sw/device/lib/base:mmio_on_device_do_not_use_directly",
+        "//sw/device/lib/base:mmio_on_host_do_not_use_directly",
+        "//sw/device/silicon_creator/rom_ext/...",
+        "//sw/device/silicon_creator/rom/keys/real:real",
+    ]
+    DEVICE_LIBS_QUERY = (f"kind(cc_library, ({' + '.join(DEVICE_LIBS_INC)}) "
+                         f"- ({' + '.join(DEVICE_LIBS_EXC)}))")
+
+    # Gather and build all device libs that can be built on host.
+    # We handle everyting in a single cquery instead of one cquery per test target since
+    # each invocation incurs an additional startup cost.
+    device_libs_all = run(BAZEL, "cquery", DEVICE_LIBS_QUERY,
+                          "--output=starlark", "--starlark:expr=target.label")
+    device_libs_incompat = run(
+        BAZEL, "cquery", DEVICE_LIBS_QUERY, "--output=starlark",
+        ("--starlark:expr=target.label "
+         "if 'IncompatiblePlatformProvider' in providers(target) "
+         "else ''"))
+    log.info(f"incompatible libraries: {pformat(device_libs_incompat)}")
+    device_libs = list(set(device_libs_all) - set(device_libs_incompat))
+    log.info(f"instrumented libraries: {pformat(device_libs)}")
+    # Force rebuild using a random action_env variable to make sure we have all the
+    # object (*.o) files that we want.
+    run(BAZEL, "build", "--config=coverage_clang",
+        f"--action_env=REBUILD={randint(0, 1000)}", *device_libs)
```

Comment:
I think you're right. I added this because I was having issues when I was trying different versions of `clang` and it was faster than `bazel clean --expunge`. IIRC `bazel` was thinking that everything was up to date even though I switched from clang-12 to clang-13 for example. I think we can remove this now.